Promote volume limits to GA

bertinatto opened the pull request:
**What type of PR is this?**

Graduates volumes limits to GA.

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Node-specific volume limits has graduated to GA.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/sig storage

bertinatto commented:
The next step for this PR is to enable the e2e tests we currently have for volume limits.

I initially tried with the GCE PD driver, but its limit is 127 and the test takes a lot of time to finish.

Following @msau42's suggestion I distributed all the volumes among 10 pods, but it didn't improve much either. Here are some numbers (with a custom limit of 100 volumes):

10 pods, 100 volumes: 2237.883 seconds
1 pod, 100 volumes: 1827.622 seconds

bertinatto commented:
I did another benchmark trying to schedule 1 pod per volume.

Somewhere after scheduling 62 pods, the test started to get rate-limited. I aborted after 3 hours, and by then it scheduled 73 pods.

bertinatto committed:
Promote volume limits to GA

bertinatto commented:
Results from job `pull-kubernetes-e2e-gce-csi-serial`, where 1 pod per volume was scheduled:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/83568/pull-kubernetes-e2e-gce-csi-serial/1184776157791260672/build-log.txt

```
I1017 12:21:34.466] [1mSTEP[0m: Waiting for the pods to be Running
W1017 12:21:55.307] 2019/10/17 12:21:55 process.go:199: Interrupt after 2h0m0s timeout during ./hack/ginkgo-e2e.sh --ginkgo.focus=CSI.*(\[Serial\]|\[Disruptive\]) --ginkgo.skip=\[Flaky\]|\[Feature:.+\]|\[Slow\] --minStartupPods=8 --report-dir=/workspace/_artifacts --disable-log-dump=true. Will terminate in another 15m
I1017 12:21:55.409] Oct 17 12:21:55.308: INFO: Running AfterSuite actions on all nodes
```

In the beginning pods were taking around 20 seconds to get `Running`, then they were taking around 9 seconds.

bertinatto commented:
/test pull-kubernetes-e2e-gce-csi-serial

bertinatto commented:
/test pull-kubernetes-e2e-gce

msau42 commented:
I think with 1 pod per volume we will hit cpu resource limits per node (as well as the 100 pods per node).  How about something like 5-10 volumes per pod?

msau42 commented:
/assign @davidz627 

> 10 pods, 100 volumes: 2237.883 seconds
> 1 pod, 100 volumes: 1827.622 seconds

For the 1 pod case, 18 seconds per volume is not terrible considering gce API rate limiting and k8s exponential backoff.  But agree 30 mins is too long for a k/k test.  Even if we got it down to 10 seconds per volume, that's still 20 mins.

Maybe for k/k, we have a test override to test less. But the test becomes more of a stress test rather than testing the feature (which is also fine because it has found attacher issues in the past).  And for the actual feature testing, we rely on hostpath CSI.

Thoughts?

davidz627 commented:
> But the test becomes more of a stress test rather than testing the feature (which is also fine because it has found attacher issues in the past). And for the actual feature testing, we rely on hostpath CSI.

I would rather not have our pull job be "more of a stress test." It would be fine in a periodic maybe but having a pull job use so many resources and take so long seems undesirable (I believe gce-serial is run as pull in some circumstances).

I'm ok with adding a flag to control the Volume Limit on the PD Driver to lighten the load on the test. I can [sort of] imagine a real user use-case for this as well where an admin may want to limit the amount of PD's the users can use on one node.

msau42 commented:
We have optional gce-serial and gce-slow jobs running for any PR that touches sig-storage files, but agree that we don't want stress testing to be part of pulls. We actually have the stress single node and stress cluster wide storage scalability tests now, but we have to enable them.

bertinatto committed:
Enable volume limits e2e tests for CSI hostpath driver

bertinatto commented:
OK, so I'll test the feature using the CSI hostpath driver for now. Once https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/422 is implemented in the PD driver, we can enable the test for that driver as well.

There's another test [here](https://github.com/bertinatto/kubernetes/blob/82a866d243a319180c4576f27e6390f07264f331/test/e2e/storage/csi_mock_volume.go#L385) that also tests the feature. I might remove it in favor of the testsuite

@msau42, could you please check https://github.com/kubernetes-csi/csi-driver-host-path/pull/110 and release a new version of the driver (so I can use it here)?

bertinatto commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -109,7 +109,7 @@ func (t *volumeLimitsTestSuite) defineTests(driver TestDriver, pattern testpatte
 	// And one extra pod with a CSI volume should get Pending with a condition
 	// that says it's unschedulable because of volume limit.
 	// BEWARE: the test may create lot of volumes and it's really slow.
-	ginkgo.It("should support volume limits [Slow][Serial]", func() {
```
This will be picked up by the job `pull-kubernetes-e2e-gce-csi-serial`.

msau42 replied:
Should we still keep serial? Is it not slow anymore?

liggitt replied:
same question, isn't it still `[Slow]`?

bertinatto commented on test/e2e/storage/testsuites/volumelimits.go:
```diff
@@ -326,7 +330,7 @@ func getCSINodeLimits(cs clientset.Interface, nodeName string, driverInfo *Drive
 		var csiDriver *storagev1beta1.CSINodeDriver
 		for _, c := range csiNode.Spec.Drivers {
-			if c.Name == driverInfo.Name {
+			if c.Name == driverInfo.Name || c.Name == config.GetUniqueDriverName() {
```
Some drivers have their name updated (e.g., `csi-hostpath` to `csi-hostpath-namespace`).

bertinatto commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -257,10 +258,9 @@ func InitMockCSIDriver(driverOpts CSIMockDriverOpts) testsuites.TestDriver {
-				testsuites.CapPersistence:  false,
-				testsuites.CapFsGroup:      false,
-				testsuites.CapExec:         false,
-				testsuites.CapVolumeLimits: true,
```
The mock drivers does support volume limits, but the test isn't running for this drivers anyway (it doesn't support persistence).

liggitt replied:
I'm confused why we would drop this capability, then

bertinatto replied:
To clarify: I dropped because I found it misleading. It took me a while to understand why a test with this capability wouldn't run in any job.

bertinatto committed:
Update hostpath CSI driver manifest

This is required for promoting volume limits to GA. The new version of
the driver reports the max number of volumes it supports. Such number
should be specified as a CLI argument when starting the driver.

gnufied commented:
/priority important-soon

msau42 commented on pkg/features/kube_features.go:
```diff
@@ -282,6 +282,8 @@ const (
 
 	// owner: @gnufied
 	// beta : v1.12
+	// GA   : v1.17
```
you also need to flip the feature gate further down in this file.

msau42 commented on pkg/features/kube_features.go:
```diff
@@ -521,7 +523,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	ExpandInUsePersistentVolumes:   {Default: true, PreRelease: featuregate.Beta},
 	ExpandCSIVolumes:               {Default: true, PreRelease: featuregate.Beta},
-	AttachVolumeLimit:              {Default: true, PreRelease: featuregate.Beta},
+	AttachVolumeLimit:              {Default: true, PreRelease: featuregate.GA},
```
I believe we should also set locktodefault

bertinatto replied:
Added

msau42 commented:
@gnufied there is some concern about how to handle the shared limits scenario, where the volume limit could be shared by multiple drivers (such as scsi limit), or even other hardware devices.  Do we have a solution for those cases?

@kubernetes/sig-storage-pr-reviews

bertinatto commented:
/retest

bertinatto commented:
@msau42 I noticed we didn't update the scheduler to use the GA version of `CSINode` in https://github.com/kubernetes/kubernetes/pull/83474. Was that intentional?

msau42 commented:
> @msau42 I noticed we didn't update the scheduler to use the GA version of `CSINode` in #83474. Was that intentional?

You mentioned you were going to handle it so I didn't want to conflict since you were also making changes in the same area. Let me know if you still plan to, otherwise I can take it up.

gnufied commented:
@msau42 The current design leaves details of how to incorporate such things with CSI Driver. The idea is - if a user is deploying k8s on heterogeneous nodes then he must use daemonset with different selectors to deploy CSI driver and override the limits as they see fit.  

@justinsb @leakingtapan is that something that will work in long term?

msau42 commented:
> daemonset with different selectors to deploy CSI driver and override the limits as they see fit.

Alternatively driver overrides can be specified as percentage or fixed number to subtract so you don't need to deploy multiple daemonsets per machine type.

I think having the admin preallocate blocks is an ok compromise in this case.  I think supporting limits shared by multiple resource types in a generic way is going to be a much bigger challenge

bertinatto committed:
Convert scheduler to use CSINode GA

bertinatto commented:
/test pull-kubernetes-e2e-gce

bertinatto commented on pkg/features/kube_features.go:
```diff
@@ -522,7 +524,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	ExpandInUsePersistentVolumes:   {Default: true, PreRelease: featuregate.Beta},
 	ExpandCSIVolumes:               {Default: true, PreRelease: featuregate.Beta},
-	AttachVolumeLimit:              {Default: true, PreRelease: featuregate.Beta},
+	AttachVolumeLimit:              {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.19
```
I'm a bit confused. I noticed that we set `LockToDefault` to `true` in `CSINodeInfo` as well, but if I understand the docs [1] correctly, these features are CLI arguments and they should work for at least one version:

```
Rule #5b: CLI elements of admin-facing components (e.g. kubelet) must function after their announced deprecation for no less than:

    GA: 6 months or 1 release (whichever is longer)
```

Shouldn't these features be deprecated (keep working with a warning) in this version and only locked in 1.18?

CC @msau42 @liggitt 

[1] https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli

liggitt replied:
see feature-gate-specific lifecycle in [section 7](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-feature-or-behavior):

> An exception to the above rule is feature gates. Feature gates are key=value pairs that allow for users to enable/disable experimental features.
> 
> Feature gates are intended to cover the development life cycle of a feature - they are not intended to be long-term APIs. As such, they are expected to be deprecated and removed after a feature becomes GA or is dropped.
> 
> As a feature moves through the stages, the associated feature gate evolves. The feature life cycle matched to its corresponding feature gate is:
> ...
> GA: the feature gate is deprecated (see “Deprecation”) and becomes non-operational.

bertinatto commented:
/retest

bertinatto commented:
/test pull-kubernetes-e2e-gce-100-performance

bertinatto commented:
/test pull-kubernetes-e2e-gce-100-performance

liggitt commented:
is the "Convert scheduler to use CSINode GA" commit related to this feature? I expected that to be done in 1.18

liggitt commented on pkg/features/kube_features.go:
```diff
@@ -522,7 +524,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 	ExpandInUsePersistentVolumes:   {Default: true, PreRelease: featuregate.Beta},
 	ExpandCSIVolumes:               {Default: true, PreRelease: featuregate.Beta},
-	AttachVolumeLimit:              {Default: true, PreRelease: featuregate.Beta},
+	AttachVolumeLimit:              {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.19
```
drop all associated checks of `...FeatureGate.Enabled(features.AttachVolumeLimit)` and assume they evaluate to true now

bertinatto replied:
Done

bertinatto commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go:
```diff
@@ -498,10 +498,8 @@ func ClusterRoles() []rbacv1.ClusterRole {
-		utilfeature.DefaultFeatureGate.Enabled(features.AttachVolumeLimit) {
-		kubeSchedulerRules = append(kubeSchedulerRules, rbacv1helpers.NewRule(Read...).Groups(storageGroup).Resources("csinodes").RuleOrDie())
+		// Needed for volume limits
+		rbacv1helpers.NewRule(Read...).Groups(storageGroup).Resources("csinodes").RuleOrDie(),
```
I'm removing **both** _CSINodeInfo_ and _AttachVolumeLimits_ checks because since _AttachVolumeLimits_ is always true now, this rules needs to exist.

liggitt commented:
mechanics of this change lgtm

msau42 commented:
/lgtm
/approve

jsafrane commented:
/lgtm

liggitt commented:
/approve

bertinatto commented:
/retest
